pkg/operators/combiner/combiner.go

```diff
@@ -0,0 +1,309 @@
+	return Priority
+}
+
+type combinerConfig struct {
+	// Interval to wait for data before emitting the combined data
+	interval time.Duration
+
+	// Buffer to send data to the combiner data source
+	packetBuf chan datasource.PacketArray
+}
+
+type combinerOperatorInstance struct {
+	// Number of targets to (ideally) wait for before emitting the combined data
+	targets int
+
+	configs map[datasource.DataSource]*combinerConfig
+
+	done chan struct{}
+}
+
+func (o *combinerOperatorInstance) Name() string {
+	return OperatorName
+}
+
+func (o *combinerOperatorInstance) InstanceParams() params.ParamDescs {
+	return nil
+}
+
+func (o *combinerOperatorInstance) ExtraParams(gadgetCtx operators.GadgetContext) api.Params {
+	return nil
+}
+
+func (o *combinerOperatorInstance) forwardData(
+	gadgetCtx operators.GadgetContext,
+	config *combinerConfig,
+	combinedDs datasource.DataSource,
+) {
+	combinedPacket, err := combinedDs.NewPacketArray()
+	if err != nil {
+		gadgetCtx.Logger().Debugf("combiner: failed to create new packet array: %s", err)
+		return
+	}
+	defer func() {
+		if combinedPacket != nil {
+			combinedDs.Release(combinedPacket)
+		}
+	}()
+
+	targetCount := 0
+
+	var c <-chan time.Time
+
+	if config.interval == 0 {
+		// Define a maximum waiting time for data from all targets
+		// TODO: Make it configurable?
+		timeout := time.NewTimer(5 * time.Second)
+		defer timeout.Stop()
+		c = timeout.C
+	} else {
+		// Even if we receive data from all targets, we emit the combined data
+		// only after the requested interval
+		ticker := time.NewTicker(config.interval)
+		defer ticker.Stop()
+		c = ticker.C
+	}
+
+	emitAndAllocate := func() error {
+		if err := combinedDs.EmitAndRelease(combinedPacket); err != nil {
+			gadgetCtx.Logger().Errorf("Failed emitting data array for ds combiner %q: %s",
+				combinedDs.Name(), err)
+		}
+
+		// Allocate a new packet array for next iteration
+		combinedPacket, err = combinedDs.NewPacketArray()
+		if err != nil {
+			// We can't continue without a new packet array
+			return fmt.Errorf("creating new packet array: %w", err)
+		}
+
+		targetCount = 0
+		return nil
+	}
+
+	for {
+		select {
+		case <-o.done:
+			gadgetCtx.Logger().Debugf("combiner: done with %q", combinedDs.Name())
+			return
+		case inPacket := <-config.packetBuf:
+			targetCount++
+
+			// TODO: Support Append multiple packets at once?
+			for i := 0; i < inPacket.Len(); i++ {
+				combinedPacket.Append(inPacket.Get(i))
+			}
+
+			// For data sources that don't have an interval, we wait for data
+			// from all targets before emitting the combined data.
+			if config.interval == 0 && targetCount == o.targets {
+				if err := emitAndAllocate(); err != nil {
+					gadgetCtx.Logger().Errorf("Failed emitting and allocating combined data: %s", err)
+				}
+				return
+			}
+		case <-c:
+			if config.interval == 0 {
+				gadgetCtx.Logger().Warnf("Data is incomplete: timeout waiting for data from all targets (%d/%d)",
+					targetCount, o.targets)
```

Comment:
Don't we need an emitAndRelease() here + return?

scenario with the current code:
1. We have 3 nodes
2. We set interval=0
3. We received packages from 2 nodes -> `targetCount==2`
4. 5 second timer runs out, because third node was a little late, and we are in the `<-c` case
5. We print the warning and call `emitAndAllocate()` -> `targetCount` is set to 0
6. We are still in the `for` loop and receive now the last package, so we are in the `inPacket := <-config.packetBuf:` case
7. The last package is now grouped with the second intervals from the first 2 nodes
8. ...

If I understood correctly `interval == 0` seems to be a one shot, nothing periodic, right? The `return` in the `if config.interval == 0 && targetCount == o.targets` indicates that at least.

Reply:
> If I understood correctly interval == 0 seems to be a one shot, nothing periodic, right? The return in the if config.interval == 0 && targetCount == o.targets indicates that at least.
>
Correct.

> 7. The last package is now grouped with the second intervals from the first 2 nodes
>
There's no second interval because `interval` is 0. So, yes, the packet with be stored to be emitted, but it never gets actually emitted. However, you're right. For cleanness, it should return but the outcome will not change.